STORM-561: Add flux as an external module

ptgoetz opened the pull request:
For a description of everything this does, it's probably easiest to look at the README file:

https://github.com/ptgoetz/storm/blob/STORM-561/external/flux/README.md

HeartSaVioR commented on external/flux/README.md:
````diff
@@ -0,0 +1,834 @@
+The current version of Flux is available in Maven Central at the following coordinates:
+```xml
+<dependency>
+    <groupId>com.github.ptgoetz</groupId>
````
It should be changed to fit to Storm external.

HeartSaVioR commented on external/flux/README.md:
```diff
@@ -0,0 +1,834 @@
+<dependencies>
+    <!-- Flux include -->
+    <dependency>
+        <groupId>com.github.ptgoetz</groupId>
```
Once again, it should be changed to fit to Storm external.

ptgoetz commented:
@HeartSaVioR,

Thanks for the review!

Thanks for catching that... I will fix it.

HeartSaVioR commented on external/flux/flux-core/pom.xml:
```diff
@@ -0,0 +1,93 @@
+    <packaging>jar</packaging>
+
+    <name>flux-core</name>
+    <url>https://github.com/ptgoetz/flux</url>
```
Maybe removing url is better to describe it is Storm's external now.

HeartSaVioR commented on external/flux/flux-examples/README.md:
```diff
@@ -0,0 +1,68 @@
+# Flux Examples
+A collection of examples illustrating various capabilities.
+
+## Building From Source and Running
```
It should be changed to clone Storm project and cd to external/flux, and build. (or build whole modules from Storm root)

HeartSaVioR commented on external/flux/flux-examples/pom.xml:
```diff
@@ -0,0 +1,86 @@
+    <packaging>jar</packaging>
+
+    <name>flux-examples</name>
+    <url>https://github.com/ptgoetz/flux</url>
```
Maybe removing url is better to describe it is Storm's external now.

ptgoetz commented:
Thanks again for the review @HeartSaVioR. I've addressed all your comments.

HeartSaVioR commented on external/flux/README.md:
```diff
@@ -0,0 +1,834 @@
+                              build, validate, and print information about
+                              the topology.
+ -e,--env-filter              Perform environment variable substitution.
+                              Replace keysidentified with `${ENV-[NAME]}`
```
keysidentified => keys identified (needs space)

HeartSaVioR commented:
I've looked at documentations of flux and leave comments, and ran unit tests and all passed.
Since this introduces about to 6000 lines of code, I wanna check functionalities first instead of reading whole codes.
When I have some time, I'll try this on storm-starter and report any issues.

As a users of Storm, thanks for introducing really amazing feature! 
It'll really makes Storm easy to configure.

HeartSaVioR commented:
I'm trying flux-example first, and found multilang.yaml doesn't really work cause it doesn't include randomsentence.js and splitsentence.py 
Users may expect runnable examples, so it would be better to copy two files from storm-starter and place them to src/main/resources/resources, or place them to multilang/resources and modify pom.xml although it's duplicated to storm-starter.

HeartSaVioR commented:
Same reason it may be better to add dependencies related to storm-kafka to flux-example to make kafka example run properly, or discard kafka example since hdfs/hbase example covers flux features enough.

HeartSaVioR commented on external/flux/flux-examples/README.md:
```diff
@@ -0,0 +1,68 @@
+
+This example illustrates how to use Flux to setup a storm-hbase bolt to write to HBase.
+
+In order to use this example, you will need to edit the `src/main resrouces/hbase-site.xml` file to reflect your HBase
```
You may want to change how to run as same as simple_hdfs.yaml since you prepared hbase_bolt.properties and simple_hbase.yaml is ready to reflect.

HeartSaVioR replied:
@ptgoetz Just a reminder, you may want to check it since it contains typo (resrouces) and hbase-site.xml file is not on directory. And actually we can run simple_hbase.yaml with similar steps from simple_hdfs.yaml.

HeartSaVioR commented:
I follow simple_hbase test with using properties filter, and failed to run.

Running with --local it prints 
```
Received topology submission for hbase-persistent-wordcount with conf {"topology.max.task.parallelism" nil, "topology.acker.executors" nil, "topology.kryo.register" nil, "topology.kryo.decorators" (), "topology.name" "hbase-persistent-wordcount", "storm.id" "hbase-persistent-wordcount-1-1431039734", "hbase.conf" {"hbase.rootdir" "hdfs://<hdfs>:8020/hbase", "hbase.zookeeper.quorum" "<zk>:2181"}, "topology.workers" 1}
```
But HBaseBolt throws TableNotFoundException, 
```
7124 [Thread-11-hbase-bolt] WARN  org.apache.hadoop.hbase.client.HConnectionManager$HConnectionImplementation - Encountered problems when prefetch hbase:meta table:
org.apache.hadoop.hbase.TableNotFoundException: Cannot find row in hbase:meta for table: WordCount, row=WordCount,nathan,99999999999999
	at org.apache.hadoop.hbase.client.MetaScanner.metaScan(MetaScanner.java:146) ~[flux-examples-0.11.0-SNAPSHOT.jar:0.11.0-SNAPSHOT]
	at org.apache.hadoop.hbase.client.HConnectionManager$HConnectionImplementation.prefetchRegionCache(HConnectionManager.java:1159) [flux-examples-0.11.0-SNAPSHOT.jar:0.11.0-SNAPSHOT]
	at org.apache.hadoop.hbase.client.HConnectionManager$HConnectionImplementation.locateRegionInMeta(HConnectionManager.java:1223) [flux-examples-0.11.0-SNAPSHOT.jar:0.11.0-SNAPSHOT]
	at org.apache.hadoop.hbase.client.HConnectionManager$HConnectionImplementation.locateRegion(HConnectionManager.java:1111) [flux-examples-0.11.0-SNAPSHOT.jar:0.11.0-SNAPSHOT]
	at org.apache.hadoop.hbase.client.HConnectionManager$HConnectionImplementation.locateRegion(HConnectionManager.java:1068) [flux-examples-0.11.0-SNAPSHOT.jar:0.11.0-SNAPSHOT]
	at org.apache.hadoop.hbase.client.AsyncProcess.findDestLocation(AsyncProcess.java:365) [flux-examples-0.11.0-SNAPSHOT.jar:0.11.0-SNAPSHOT]
	at org.apache.hadoop.hbase.client.AsyncProcess.submit(AsyncProcess.java:507) [flux-examples-0.11.0-SNAPSHOT.jar:0.11.0-SNAPSHOT]
	at org.apache.hadoop.hbase.client.AsyncProcess.submitAll(AsyncProcess.java:476) [flux-examples-0.11.0-SNAPSHOT.jar:0.11.0-SNAPSHOT]
	at org.apache.hadoop.hbase.client.HConnectionManager$HConnectionImplementation.processBatchCallback(HConnectionManager.java:2355) [flux-examples-0.11.0-SNAPSHOT.jar:0.11.0-SNAPSHOT]
	at org.apache.hadoop.hbase.client.HTable.batchCallback(HTable.java:835) [flux-examples-0.11.0-SNAPSHOT.jar:0.11.0-SNAPSHOT]
	at org.apache.hadoop.hbase.client.HTable.batch(HTable.java:814) [flux-examples-0.11.0-SNAPSHOT.jar:0.11.0-SNAPSHOT]
	at org.apache.storm.hbase.common.HBaseClient.batchMutate(HBaseClient.java:100) [flux-examples-0.11.0-SNAPSHOT.jar:0.11.0-SNAPSHOT]
	at org.apache.storm.hbase.bolt.HBaseBolt.execute(HBaseBolt.java:63) [flux-examples-0.11.0-SNAPSHOT.jar:0.11.0-SNAPSHOT]
	at backtype.storm.daemon.executor$fn__4722$tuple_action_fn__4724.invoke(executor.clj:633) [storm-core-0.9.4.jar:0.9.4]
	at backtype.storm.daemon.executor$mk_task_receiver$fn__4645.invoke(executor.clj:401) [storm-core-0.9.4.jar:0.9.4]
	at backtype.storm.disruptor$clojure_handler$reify__1446.onEvent(disruptor.clj:58) [storm-core-0.9.4.jar:0.9.4]
	at backtype.storm.utils.DisruptorQueue.consumeBatchToCursor(DisruptorQueue.java:120) [storm-core-0.9.4.jar:0.9.4]
	at backtype.storm.utils.DisruptorQueue.consumeBatchWhenAvailable(DisruptorQueue.java:99) [storm-core-0.9.4.jar:0.9.4]
	at backtype.storm.disruptor$consume_batch_when_available.invoke(disruptor.clj:80) [storm-core-0.9.4.jar:0.9.4]
	at backtype.storm.daemon.executor$fn__4722$fn__4734$fn__4781.invoke(executor.clj:748) [storm-core-0.9.4.jar:0.9.4]
	at backtype.storm.util$async_loop$fn__458.invoke(util.clj:463) [storm-core-0.9.4.jar:0.9.4]
	at clojure.lang.AFn.run(AFn.java:24) [clojure-1.5.1.jar:na]
	at java.lang.Thread.run(Thread.java:745) [na:1.7.0_55]
```

Table is enabled and have one region.
WordCount	1	'WordCount', {NAME => 'cf'}

I didn't clarify yet it is a storm-hbase issue or flux issue.

HeartSaVioR commented:
Don't mind last comment, I was mistaken so that hdfs and hbase zk quorum are different cluster.

HeartSaVioR commented:
Ran all examples on flux-example by local mode and leave some comments.
Other things work fine.

ptgoetz commented:
@HeartSaVioR nice catch on the multilang scripts -- I accidentally deleted them when I switched over to using the new multilang packaging. They are back now.

I also added kafka as a dependency to flux-examples, and fixed a typo.

vesense commented:
Good job! We also developed a componet like FLUX. I'm happy to try.

harshach commented:
+1 looks good to me. Ran examples on local mode and wordcount on storm cluster.

gavinfish commented:
Flux is a great job! 

I want to know how can I update my topology when the configuration has been changed without restarting it, something like hot deployment in tomcat?

revans2 commented:
@gavinfish that is a very difficult thing to do correctly (Which is why we do not currently support it).  As part of the JStorm integration work we are looking at adding some support for that with STORM-1334, STORM-1335 and STORM-1330.  They should provide some functionality for what you want, but not everything.
